### PR TITLE
Add max_norm for gradient clipping.

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: Add max_norm as user options for gradient clipping

1. By Default, we use 1.0f as max_norm when gradient clipping is enabled. 
2. This change is to allow user to specify the max_norm.
3. For old ORTTrainer API, user need specify the max_norm in map_optimizer_attributes func, just as how we pass "epsilon"
4. For new ORTTrainer API, user can specify the value when creating optim.AdamConfig or optim.LambConfig

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by pengwa** [comment]
@SherlockNoMad  @wschin  I was a little bit hesitating to make max_norm for clipping as Adam/Lamb's attribute. While if we don't do this way, we need introduce another input for Adam/Lamb called max_norm, this also need a schema change. I would like to see how you think about the possible best solution. Thanks a lot!

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 1]
```diff
@@ -15,11 +15,12 @@ namespace cuda {
 template<typename TLossScale, typename TGradNorm, typename TFinalScale>
 __device__ __forceinline__ TFinalScale _ComputeGradScale(
 const TLossScale* loss_scale,
-const TGradNorm* g_norm) {
+const TGradNorm* g_norm,
+const TFinalScale max_norm) {
 TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:20]: we referece https://pytorch.org/docs/stable/_modules/torch/nn/utils/clip_grad.html#clip_grad_norm_ for the clipping.

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 2]
```diff
@@ -15,11 +15,12 @@ namespace cuda {
...
-if (g_norm != nullptr && TFinalScale(*g_norm) > scale) {
+if (g_norm != nullptr && TFinalScale(*g_norm) > TFinalScale(scale * max_norm)) {
     const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:22]: @wschin do you know why we divide g_norm with scale , got a actual_g_norm ,
 then  scale multiple the actual_g_norm back. This looks interesting, any reason we did this?

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 3]
```diff
@@ -15,11 +15,12 @@ namespace cuda {
 template<typename TLossScale, typename TGradNorm, typename TFinalScale>
 __device__ __forceinline__ TFinalScale _ComputeGradScale(
 const TLossScale* loss_scale,
-const TGradNorm* g_norm) {
+const TGradNorm* g_norm,
+const TFinalScale max_norm) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:19]: Can we rewrite these piece of code to be more readable ? 

it took us so long to understand this... let's make it less painful for the next reader

**Comment by SherlockNoMad** [comment]
> @SherlockNoMad @wschin I was a little bit hesitating to make max_norm for clipping as Adam/Lamb's attribute. While if we don't do this way, we need introduce another input for Adam/Lamb called max_norm, this also need a schema change. I would like to see how you think about the possible best solution. Thanks a lot!

I think we can extract gradient_norm operation outside of optimizer op. The original motivation for merging the gradient norm and loss scaling into optimizer is for

**Comment by SherlockNoMad** [comment]
>

**Comment by SherlockNoMad** [comment]
> @SherlockNoMad @wschin I was a little bit hesitating to make max_norm for clipping as Adam/Lamb's attribute. While if we don't do this way, we need introduce another input for Adam/Lamb called max_norm, this also need a schema change. I would like to see how you think about the possible best solution. Thanks a lot!

As we are moving towards ORTModule, let's just go with this "attribute" approach and don't need to complicated this to make this an input.

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh (hunk 2 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:22]: The actual equation here is to compute a `scalar` which may be used to multiply the gradient passed into optimizer (it's scaled gradient when loss scaling is enabled). If we don't have max gradient norm, it's
```
scalar = loss_scale
```
If max gradient norm is `N`, then we need to further divide the gradient to get `norm=1` gradient. Finally, we multiply the `norm=1` gradient by max norm to make it `norm=max norm` gradient.
```
scalar_for_clipiing = actual_gradient / max_norm
```
and
```
finaly_scalar = scalar * scalar_for_clipping
```

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:19]: Just added a comment to explain. I think what we need is adding that comment to this PR for details, not rewriting code.

**Review by wschin — DISMISSED** [review]

**Comment by pengwa** [comment]
@liqunfu  @thiagocrepaldi @BowenBao @spandantiwari  Feel free to let me know if you have any comments on the frontend side. Thanks a lot!

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 4]
```diff
@@ -12,16 +12,18 @@ namespace cuda {
...
-}
-return scale;
+    const TLossScale* loss_scale,
+    const TGradNorm* scaled_g_norm,
+    const TFinalScale max_norm) {
+  TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
+  TFinalScale scaled_max_norm = TFinalScale(scale * max_norm);
+  TFinalScale scaled_clipping_factor = scale;
+  if (scaled_g_norm != nullptr && TFinalScale(*scaled_g_norm) > scaled_max_norm) {
+    scaled_clipping_factor = TFinalScale(*scaled_g_norm) / (max_norm + TFinalScale(1e-6f));
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:24]: Let's assert `max_norm != 0` outside this function and remove `1e-6` here.

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 5]
```diff
@@ -12,16 +12,18 @@ namespace cuda {
...
-    const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
-    scale *= actual_g_norm;
-}
-return scale;
+    const TLossScale* loss_scale,
+    const TGradNorm* scaled_g_norm,
+    const TFinalScale max_norm) {
+  TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
+  TFinalScale scaled_max_norm = TFinalScale(scale * max_norm);
+  TFinalScale scaled_clipping_factor = scale;
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:22]:
```suggestion
  // This number is used to divide the scaled gradient before applying optimizers.
  // Note that scaled gradient is an optimizer input.
  TFinalScale scaled_g_scaling_factor = scale;
```

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 6]
```diff
@@ -12,16 +12,18 @@ namespace cuda {
...
-if (g_norm != nullptr && TFinalScale(*g_norm) > scale) {
-    const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
-    scale *= actual_g_norm;
-}
-return scale;
+    const TLossScale* loss_scale,
+    const TGradNorm* scaled_g_norm,
+    const TFinalScale max_norm) {
+  TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
+  TFinalScale scaled_max_norm = TFinalScale(scale * max_norm);
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:21]:
```suggestion
  const TFinalScale scaled_max_norm = TFinalScale(scale * max_norm);
```

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 7]
```diff
@@ -12,16 +12,18 @@ namespace cuda {
...
-TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
-if (g_norm != nullptr && TFinalScale(*g_norm) > scale) {
-    const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
-    scale *= actual_g_norm;
-}
-return scale;
+    const TLossScale* loss_scale,
+    const TGradNorm* scaled_g_norm,
+    const TFinalScale max_norm) {
+  TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:20]:
```suggestion
  const TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
```

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 8]
```diff
@@ -12,16 +12,18 @@ namespace cuda {
...
-const TGradNorm* g_norm) {
-TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
-if (g_norm != nullptr && TFinalScale(*g_norm) > scale) {
-    const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
-    scale *= actual_g_norm;
-}
-return scale;
+    const TLossScale* loss_scale,
+    const TGradNorm* scaled_g_norm,
+    const TFinalScale max_norm) {
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:19]:
```suggestion
    const TFinalScale max_g_norm) {
```

File: orttraining/orttraining/training_ops/cuda/optimizer/common.cuh [hunk 9]
```diff
@@ -12,16 +12,18 @@ namespace cuda {
...
 __device__ __forceinline__ TFinalScale _ComputeGradScale(
-const TLossScale* loss_scale,
-const TGradNorm* g_norm) {
-TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
-if (g_norm != nullptr && TFinalScale(*g_norm) > scale) {
-    const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
-    scale *= actual_g_norm;
-}
-return scale;
+    const TLossScale* loss_scale,
```
Comment (wschin) [inline orttraining/orttraining/training_ops/cuda/optimizer/common.cuh:17]:
```suggestion
    // Scale of the gradient (called "g_scale" in some comments below)
    const TLossScale* loss_scale,
```

**Review by wschin — APPROVED** [review]
:shipit:

**Comment by pengwa** [comment]
> @liqunfu @thiagocrepaldi @BowenBao @spandantiwari Feel free to let me know if you have any comments on the frontend side. Thanks a lot!

Seems need your approval to get this merged. Could you please take a quick look at the front end part, which should be a pretty straightforward change.

**Review by SherlockNoMad — DISMISSED** [review]

**Review by liqunfu — DISMISSED** [review]

**Review by thiagocrepaldi — COMMENTED** [review]
Could you add a frontend unit test that exercises this new optim parameter?

We can't be sure whether the `optim/config.py` works or guard it from regression without testing it

**Comment by pengwa** [comment]
> Could you add a frontend unit test that exercises this new optim parameter?
> 
> We can't be sure whether the `optim/config.py` works or guard it from regression without testing it

updated, please take a look.

**Review by thiagocrepaldi — APPROVED** [review]
